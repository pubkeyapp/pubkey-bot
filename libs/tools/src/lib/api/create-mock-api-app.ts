import { Tree } from '@nx/devkit'
import { applicationGenerator, libraryGenerator, serviceGenerator } from '@nx/nest'

export async function createMockApiApp(tree: Tree, app: string) {
  // Build the mock app and core libs
  await applicationGenerator(tree, { name: app, directory: 'apps', skipFormat: true })
  // Create the core data access lib
  await libraryGenerator(tree, { name: `data-access`, directory: `libs/${app}/core`, skipFormat: true })
  // Create the core service
  await serviceGenerator(tree, {
    directory: 'lib',
    flat: true,
    name: `${app}-core`,
    project: `${app}-core-data-access`,
    skipFormat: true,
  })
  // Create the core feature lib
  await libraryGenerator(tree, { name: `feature`, directory: `libs/${app}/core`, skipFormat: true })
}
