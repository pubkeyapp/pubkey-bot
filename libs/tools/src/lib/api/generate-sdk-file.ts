import { names, readProjectConfiguration, Tree } from '@nx/devkit'
import { NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'

export async function generateSdkFile(tree: Tree, options: NormalizedApiFeatureSchema) {
  const project = readProjectConfiguration(tree, `sdk`)

  if (!project) {
    throw new Error(`Could not find project: ${project}`)
  }

  const sdkRoot = `${project.sourceRoot}/graphql`
  const sdkFile = `feature-${options.name}.graphql`

  const exists = tree
    .children(sdkRoot)
    .filter((file) => file.endsWith('.graphql'))
    .includes(sdkFile)

  if (!exists) {
    tree.write(`${sdkRoot}/${sdkFile}`, sdkTemplate(options.name, options.label))
  }
}

function sdkTemplate(name: string, label: string) {
  const { className, propertyName } = names(name)
  return `fragment ${className}Details on ${className} {
  createdAt
  id
  ${label}
  updatedAt
}

query adminFindMany${className}($input: AdminFindMany${className}Input!) {
  paging: adminFindMany${className}(input: $input) {
    data {
      ...${className}Details
    }
    meta {
      ...PagingMetaDetails
    }
  }
}

query adminFindOne${className}($${propertyName}Id: String!) {
  item: adminFindOne${className}(${propertyName}Id: $${propertyName}Id) {
    ...${className}Details
  }
}

mutation adminCreate${className}($input: AdminCreate${className}Input!) {
  created: adminCreate${className}(input: $input) {
    ...${className}Details
  }
}

mutation adminUpdate${className}($${propertyName}Id: String!, $input: AdminUpdate${className}Input!) {
  updated: adminUpdate${className}(${propertyName}Id: $${propertyName}Id, input: $input) {
    ...${className}Details
  }
}

mutation adminDelete${className}($${propertyName}Id: String!) {
  deleted: adminDelete${className}(${propertyName}Id: $${propertyName}Id)
}
`
}
