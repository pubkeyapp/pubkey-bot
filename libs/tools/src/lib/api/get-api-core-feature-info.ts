import { readProjectConfiguration, Tree } from '@nx/devkit'

export function getApiCoreFeatureInfo(tree: Tree, app: string) {
  const project = readProjectConfiguration(tree, `${app}-core-feature`)

  const coreFeatureModulePath = `${project.sourceRoot}/lib/${project.name}.module.ts`

  if (!tree.exists(coreFeatureModulePath)) {
    throw new Error(`getApiCoreFeatureInfo: ${coreFeatureModulePath} does not exist in ${project.sourceRoot}`)
  }

  return { coreFeatureModulePath }
}
