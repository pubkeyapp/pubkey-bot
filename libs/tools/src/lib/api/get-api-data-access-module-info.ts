import { readProjectConfiguration, Tree } from '@nx/devkit'
import { NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'

export function getApiDataAccessModuleInfo(tree: Tree, options: NormalizedApiFeatureSchema) {
  const project = readProjectConfiguration(tree, `${options.app}-${options.name}-data-access`)

  const dataAccessProjectRoot = project.sourceRoot
  const dataAccessModulePath = `${dataAccessProjectRoot}/lib/${project.name}.module.ts`

  return {
    dataAccessProjectRoot,
    dataAccessModulePath,
  }
}
