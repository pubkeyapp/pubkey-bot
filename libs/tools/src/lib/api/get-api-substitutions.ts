import { names } from '@nx/devkit'
import { NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'

export function getApiSubstitutions(options: NormalizedApiFeatureSchema) {
  const admin = names('admin')
  const model = names(options.modelName)
  const user = names('user')

  return {
    admin,
    adminCrud: !options.skipAdminCrud,
    adminFileName: admin.fileName,
    modelFileName: model.fileName,
    app: names(options.app),
    label: names(options.label),
    model,
    npmScope: options.npmScope,
    user,
    // FIXME: Support user crud
    userCrud: false,
  }
}
