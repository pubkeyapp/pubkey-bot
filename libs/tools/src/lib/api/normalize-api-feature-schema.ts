import { Tree } from '@nx/devkit'
import { ApiFeatureGeneratorSchema, NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'

export function normalizeApiFeatureSchema(tree: Tree, schema: ApiFeatureGeneratorSchema): NormalizedApiFeatureSchema {
  const modelName = schema.name
  const npmScope = tree.read('nx.json', 'utf-8')?.match(/"npmScope": "(.*)"/)?.[1]

  return {
    app: schema.app ?? 'api',
    name: schema.name,
    label: schema.label ?? 'name',
    modelName,
    npmScope,
    skipAdminCrud: schema.skipAdminCrud ?? false,
    skipDataAccess: schema.skipDataAccess ?? false,
    skipFeature: schema.skipFeature ?? false,
    skipSdk: schema.skipSdk ?? false,
    skipUtil: schema.skipUtil ?? true,
  }
}
