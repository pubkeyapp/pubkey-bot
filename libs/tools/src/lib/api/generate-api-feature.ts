import { Tree } from '@nx/devkit'
import { NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'
import { generateApiLib } from './generate-api-lib'
import { generateSdkFile } from './generate-sdk-file'

export async function generateApiFeature(tree: Tree, options: NormalizedApiFeatureSchema) {
  // FIXME: Make sure we can run this feature twice. If the libraries already exist we should skip them
  if (!options.skipDataAccess) {
    await generateApiLib(tree, 'data-access', options)
  }
  if (!options.skipFeature) {
    await generateApiLib(tree, 'feature', options)
  }
  if (!options.skipUtil) {
    await generateApiLib(tree, 'util', options)
  }
  if (!options.skipSdk) {
    await generateSdkFile(tree, options)
  }
}
