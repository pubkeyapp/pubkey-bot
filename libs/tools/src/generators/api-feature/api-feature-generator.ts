import { formatFiles, Tree } from '@nx/devkit'
import { generateApiFeature } from '../../lib/api/generate-api-feature'
import { normalizeApiFeatureSchema } from '../../lib/api/normalize-api-feature-schema'
import { ensureNxProjectExists } from '../../lib/utils/ensure-nx-project-exists'
import { ApiFeatureGeneratorSchema } from './api-feature-schema'

export async function apiFeatureGenerator(tree: Tree, rawOptions: ApiFeatureGeneratorSchema) {
  const options = normalizeApiFeatureSchema(tree, rawOptions)
  ensureNxProjectExists(tree, options.app)
  await generateApiFeature(tree, options)
  await formatFiles(tree)
}

export default apiFeatureGenerator
