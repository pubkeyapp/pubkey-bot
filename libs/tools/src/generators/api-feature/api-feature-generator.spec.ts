import { readProjectConfiguration, Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { createMockApiApp } from '../../lib/api/create-mock-api-app'

import { apiFeatureGenerator } from './api-feature-generator'
import { ApiFeatureGeneratorSchema } from './api-feature-schema'

describe('api-feature generator', () => {
  let tree: Tree
  const options: ApiFeatureGeneratorSchema = { app: 'api', name: 'test', label: 'name', skipSdk: true }

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it('should generate the feature libraries', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    await apiFeatureGenerator(tree, options)

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.name}-${lib}`)
      expect(config).toBeDefined()
    })

    const basePathDataAccess = `libs/${options.app}/${options.name}/data-access/src`
    const basePathFeature = `libs/${options.app}/${options.name}/feature/src`

    const files = [
      `${basePathDataAccess}/index.ts`,
      `${basePathDataAccess}/lib/${options.app}-${options.name}-admin.service.ts`,
      `${basePathDataAccess}/lib/${options.app}-${options.name}-data-access.module.ts`,
      `${basePathDataAccess}/lib/${options.app}-${options.name}.service.ts`,
      `${basePathDataAccess}/lib/dto/admin-create-${options.name}.input.ts`,
      `${basePathDataAccess}/lib/dto/admin-find-many-${options.name}.input.ts`,
      `${basePathDataAccess}/lib/dto/admin-update-${options.name}.input.ts`,
      `${basePathDataAccess}/lib/helpers/get-admin-${options.name}-where.input.ts`,
      `${basePathDataAccess}/lib/entity/${options.name}.entity.ts`,
      `${basePathDataAccess}/lib/entity/${options.name}-paging.entity.ts`,
      `${basePathFeature}/index.ts`,
      `${basePathFeature}/lib/${options.app}-${options.name}-admin.resolver.ts`,
      `${basePathFeature}/lib/${options.app}-${options.name}-feature.module.ts`,
    ]

    files.forEach((file) => {
      expect(tree.exists(file)).toBeTruthy()
      expect(tree.read(file).toString()).toMatchSnapshot()
    })
  })

  it('should generate the feature libraries with util lib', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature', 'util']
    await apiFeatureGenerator(tree, { ...options, skipUtil: false })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.name}-${lib}`)
      expect(config).toBeDefined()
    })
  })

  it('should generate the feature with different name', async () => {
    const testOptions = { ...options, name: 'company' }
    await createMockApiApp(tree, testOptions.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature', 'util']
    await apiFeatureGenerator(tree, { ...testOptions, skipUtil: false })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${testOptions.app}-${testOptions.name}-${lib}`)
      expect(config).toBeDefined()
    })
  })

  it('should generate the feature libraries with custom names', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    const customName = 'custom'
    await apiFeatureGenerator(tree, { ...options, name: customName })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${customName}-${lib}`)
      expect(config).toBeDefined()
    })
  })
})
