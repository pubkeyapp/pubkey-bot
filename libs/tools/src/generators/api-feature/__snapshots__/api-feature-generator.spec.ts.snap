// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-feature generator should generate the feature libraries 1`] = `
"export * from './lib/api-test-data-access.module';
export * from './lib/api-test.service';
export * from './lib/entity/test.entity';
export * from './lib/entity/test-paging.entity';
export * from './lib/dto/admin-create-test.input';
export * from './lib/dto/admin-find-many-test.input';
export * from './lib/dto/admin-update-test.input';
"
`;

exports[`api-feature generator should generate the feature libraries 2`] = `
"import { Injectable } from '@nestjs/common';
import { ApiCoreService } from '@proj/api/core/data-access';
import { AdminCreateTestInput } from './dto/admin-create-test.input';
import { AdminFindManyTestInput } from './dto/admin-find-many-test.input';
import { AdminUpdateTestInput } from './dto/admin-update-test.input';
import { TestPaging } from './entity/test-paging.entity';
import { getAdminTestWhereInput } from './helpers/get-admin-test-where.input';

@Injectable()
export class ApiTestAdminService {
  constructor(private readonly core: ApiCoreService) {}

  async createTest(input: AdminCreateTestInput) {
    return this.core.data.test.create({ data: input });
  }

  async deleteTest(testId: string) {
    const deleted = await this.core.data.test.delete({ where: { id: testId } });
    return !!deleted;
  }

  async findManyTest(input: AdminFindManyTestInput): Promise<TestPaging> {
    return this.core.data.test
      .paginate({
        orderBy: { name: 'asc' },
        where: getAdminTestWhereInput(input),
      })
      .withPages({ limit: input.limit, page: input.page })
      .then(([data, meta]) => ({ data, meta }));
  }

  async findOneTest(testId: string) {
    return this.core.data.test.findUnique({ where: { id: testId } });
  }

  async updateTest(testId: string, input: AdminUpdateTestInput) {
    return this.core.data.test.update({ where: { id: testId }, data: input });
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries 3`] = `
"import { Module } from '@nestjs/common';
import { ApiCoreDataAccessModule } from '@proj/api/core/data-access';
import { ApiTestAdminService } from './api-test-admin.service';

import { ApiTestService } from './api-test.service';

@Module({
  imports: [ApiCoreDataAccessModule],
  providers: [ApiTestService, ApiTestAdminService],
  exports: [ApiTestService],
})
export class ApiTestDataAccessModule {}
"
`;

exports[`api-feature generator should generate the feature libraries 4`] = `
"import { Injectable } from '@nestjs/common';
import { ApiCoreService } from '@proj/api/core/data-access';
import { ApiTestAdminService } from './api-test-admin.service';

@Injectable()
export class ApiTestService {
  constructor(
    private readonly core: ApiCoreService,
    readonly admin: ApiTestAdminService
  ) {}
}
"
`;

exports[`api-feature generator should generate the feature libraries 5`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class AdminCreateTestInput {
  @Field()
  name!: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 6`] = `
"import { Field, InputType, Int } from '@nestjs/graphql';
import { PagingInput } from '@proj/api/core/data-access';

@InputType()
export class AdminFindManyTestInput extends PagingInput() {
  @Field({ nullable: true })
  search?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 7`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class AdminUpdateTestInput {
  @Field({ nullable: true })
  name?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 8`] = `
"import { Prisma } from '@prisma/client';
import { AdminFindManyTestInput } from '../dto/admin-find-many-test.input';

export function getAdminTestWhereInput(
  input: AdminFindManyTestInput
): Prisma.TestWhereInput {
  const where: Prisma.TestWhereInput = {};

  if (input.search) {
    where.OR = [
      { id: { contains: input.search, mode: 'insensitive' } },
      { name: { contains: input.search, mode: 'insensitive' } },
    ];
  }

  return where;
}
"
`;

exports[`api-feature generator should generate the feature libraries 9`] = `
"import { Field, ObjectType } from '@nestjs/graphql';

@ObjectType()
export class Test {
  @Field()
  id!: string;
  @Field({ nullable: true })
  createdAt?: Date;
  @Field({ nullable: true })
  updatedAt?: Date;
  @Field()
  name!: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 10`] = `
"import { ObjectType } from '@nestjs/graphql';
import { PagingResponse } from '@proj/api/core/data-access';
import { Test } from './test.entity';

@ObjectType()
export class TestPaging extends PagingResponse<Test>(Test) {}
"
`;

exports[`api-feature generator should generate the feature libraries 11`] = `
"export * from './lib/api-test-feature.module';
"
`;

exports[`api-feature generator should generate the feature libraries 12`] = `
"import { Resolver } from '@nestjs/graphql';
import { ApiTestService } from '@proj/api/test/data-access';
import { ApiAuthGraphQLAdminGuard } from '@proj/api/auth/data-access';
import { Mutation, Query, Args } from '@nestjs/graphql';
import { UseGuards } from '@nestjs/common';
import {
  AdminCreateTestInput,
  AdminFindManyTestInput,
  Test,
  TestPaging,
  AdminUpdateTestInput,
} from '@proj/api/test/data-access';

@Resolver()
@UseGuards(ApiAuthGraphQLAdminGuard)
export class ApiTestAdminResolver {
  constructor(private readonly service: ApiTestService) {}

  @Mutation(() => Test, { nullable: true })
  adminCreateTest(@Args('input') input: AdminCreateTestInput) {
    return this.service.admin.createTest(input);
  }

  @Mutation(() => Boolean, { nullable: true })
  adminDeleteTest(@Args('testId') testId: string) {
    return this.service.admin.deleteTest(testId);
  }

  @Query(() => TestPaging, { nullable: true })
  adminFindManyTest(@Args('input') input: AdminFindManyTestInput) {
    return this.service.admin.findManyTest(input);
  }

  @Query(() => Test, { nullable: true })
  adminFindOneTest(@Args('testId') testId: string) {
    return this.service.admin.findOneTest(testId);
  }

  @Mutation(() => Test, { nullable: true })
  adminUpdateTest(
    @Args('testId') testId: string,
    @Args('input') input: AdminUpdateTestInput
  ) {
    return this.service.admin.updateTest(testId, input);
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries 13`] = `
"import { Module } from '@nestjs/common';
import { ApiTestDataAccessModule } from '@proj/api/test/data-access';
import { ApiTestAdminResolver } from './api-test-admin.resolver';

@Module({
  imports: [ApiTestDataAccessModule],
  providers: [ApiTestAdminResolver],
})
export class ApiTestFeatureModule {}
"
`;
